software/firmware/europi.py

```diff
@@ -443,10 +451,31 @@ def centre_text(self, text):
 
         padding_top = (self.height - (len(lines) * 9)) / 2
         for index, content in enumerate(lines):
-            x_offset = int((self.width - ((len(content) + 1) * 7)) / 2) - 1
             y_offset = int((index * 9) + padding_top) - 1
-            self.text(content, x_offset, y_offset)
-        oled.show()
+            self.centred_text_line(content, y_offset)
+        self.show()
+
+    def centred_text_line(self, text, y, colour=1):
+        """Displays the given text horizontally centered on its line."""
+        x_offset = int((self.width - ((len(text) + 1) * 7)) / 2) - 1
+        self.text(text, x_offset, y, colour)
+
+    def arrow(self, x, y, direction=None, size=4, colour=1):

```

Comment:
[discussion] I'm not convinced that this method belongs in the firmware yet. The biggest indication is that there is no code in the repo that calls it. I think that we would like to see a method like this used in a couple scripts before it gets 'pulled up' into the firmware.

Reply:
This is a method I'm using in the script I'm currently working on.
I can leave that function in my script but it seems pretty generic and could be useful for others.

My thinking is that we have a Display class  containing a single method `centre_text()`.
From my understanding, this class is supposed to contain all the handy display snippets, otherwise it's just a wrapper of the `SSD1306_I2C` class with a single method `centre_text()` only used in the firmware and the calibration script.

To me, the more useful snippets we add in here, the best it is as it could give ideas and material to the contributors to build a nice UI and instead of duplicating code all around, we could enforce a unity, in the code as well as visually.

Reply:
> This is a method I'm using in the script I'm currently working on. I can leave that function in my script but it seems pretty generic and could be useful for others.
> To me, the more useful snippets we add in here, the best it is as it could give ideas and material to the contributors to build a nice UI and instead of duplicating code all around, we could enforce a unity, in the code as well as visually.


> [discussion] I'm not convinced that this method belongs in the firmware yet. 


Could this potentially be solved by making the ui.py file the place to go for things like this? I think it would much more sense to users (and be much easier to parse to see what ui things they have access to) to have one file dedicated to the UI and the firmware be for the rest of the hardware.
Perhaps moving centre_text into ui.py is a bit far given how many scripts use it, although they could all be updated fairly easily, but future things like arrow would be great to put into a separate file

Reply:
Are you thinking of adding those methods in the existing `Menu` class of `ui.py` or would you move them in a new UI class?
To me it wouldn't really make sense to add them in the `Menu` class since they can be used in any display context, not only in a menu so having a new `UI` class could make sense but I'm not entirely sure it's the best solution.

As I see it, the `Display` class is the class containing all display related methods since it's wrapping `SSD1306_I2C` and that's what `SSD1306_I2C` is.

The `Display` class is a wrapper around the `SSD1306_I2C` class which already contains similar methods (line, rect, blit etc.) so to me it makes sense to add those new methods to this wrapper class.
If you both think it should be somewhere else then let's build a UI library which would contain these 2 new methods and would wrap the exiting `SSD1306_I2C` methods (because all UI methods should be available in one single place) but then there would be 2 possible ways to display a line for instance, it would be more confusing in my opinion.

Reply:
@atulrnt I think that we are on the same page with where these should eventually end up. They would make sense as either `Display` class methods, or at least functions in a UI package. We are just not ready to add them to the firmware currently. The process to add this kind of stuff is fairly new, so please bear with us as we try to find the best process together.

I should have thought to suggest this at the outset, but we do have a firmware package that was made for this exact situation called [`experimental`](https://github.com/Allen-Synthesis/EuroPi/blob/main/software/firmware/experimental/__init__.py). At this point, I think that it would be appropriate to add both the `arrow()` and `centred_text_line` functions to a new ui module in the experimental package, and avoid editing the `Display` class yet.

Another alternative would be to keep them local to your script, as you mentioned previously, but I agree that the spirit of these methods is that they have a general use. So a more public place makes more sense to me. It just depends on how much work you want to take on.

Finally, it might make the most sense to make the above changes in the PR that introduces the script that uses these functions. This would make it easier for us reviewers to see exactly how they are used in practice, as well as make each of these PRs more cohesive. It would also unblock this PR, as I think that the high/low changes are ready.

Reply:
I completely agree with moving these functions in the experimental package, that makes a lot of sense.
I'll get in touch with you on discord to figure out how to structure this experimental UI module, for now I'll remove these methods from this PR so that we can focus on merging the HIGH and LOW constants.

Reply:
Display features have been removed.